subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy

```diff
@@ -0,0 +1,330 @@
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            CompletableFuture.runAsync({
+                println "backgroundJobExecutor.isInBackground = \${requiredServices.backgroundJobExecutor.isInBackground()}"
+            }, requiredServices.backgroundJobExecutor).exceptionally { e ->
+                println "backgroundJobExecutor.status = \${e.cause.message}"
+            }
+
+            task check {}
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        outputContains("backgroundJobExecutor.isInBackground = true")
+    }
+
+    def "buildscript execution is not on the background executor"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            println "backgroundJobExecutor.isInBackground = \${requiredServices.backgroundJobExecutor.isInBackground()}"
+
+            task check {}
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        outputContains("backgroundJobExecutor.isInBackground = false")
+    }
+
+    def "task execution is not on the background executor"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+
+            task check {
+                doLast {
+                    println "backgroundJobExecutor.isInBackground = \${requiredServices.backgroundJobExecutor.isInBackground()}"
+                }
+            }
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        outputContains("backgroundJobExecutor.isInBackground = false")
+
+        when:
+        succeeds("check")
+
+        then:
+        outputContains("backgroundJobExecutor.isInBackground = false")
+    }
+
+    @IgnoreIf({ !GradleContextualExecuter.configCache })
+    def "configuration inputs are not tracked for the job"() {
+        given:
+        buildFile << """
+            import ${CompletableFuture.name}
+
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            def executor = requiredServices.backgroundJobExecutor
+            def future = CompletableFuture.runAsync({
+                println "backgroundJob.property = \${System.getProperty("property")}"
+            }, executor)
+
+            // Block until the job completes to ensure it run at the configuration time.
+            future.get()
+
+            task check {}
+        """
+
+        when:
+        succeeds("check", "-Dproperty=value")
+
+        then:
+        outputContains("backgroundJob.property = value")
+        postBuildOutputContains("Configuration cache entry stored")
```

Comment:
It would be better to use the `ConfigurationCacheFixture` rather than hard-coded console output to check for loads and stores.